libvast/src/arrow_table_slice.cpp

```diff
@@ -1277,4 +1278,155 @@ template class arrow_table_slice<fbs::table_slice::arrow::v0>;
+        die("append null value to subnet builder failed");
+    } else {
+      auto span = std::span<const uint8_t, 17>{arr->GetValue(row), 17};
+      if (!(subnet_builder->Append().ok()
+            && address_builder->Append(arr->GetValue(row)).ok()
+            && length_builder->Append(span[16]).ok()))
+        die("rebuilding of subnet struct array failed");
+    }
+  }
+  if (auto result = subnet_builder->Finish(); result.ok())
+    return *result;
+  die("failed to finish Arrow subnet column builder");
+}
+
+// TODO: this is copied from
+// experimental_table_slice_builder/enum_column_builder, can this live somewhere
+// shared?
+std::shared_ptr<arrow::Array> make_dict_array(const enumeration_type& et) {
+  arrow::StringBuilder dict_builder{};
+  for (const auto& f : et.fields())
+    if (!dict_builder.Append(std::string{f.name}).ok())
+      die("failed to build Arrow enum field array");
+  if (auto array = dict_builder.Finish(); array.ok()) {
+    return *array;
+  }
+  die("failed to finish Arrow enum field array");
+}
+
+/// Consumes arrow arrays and builds up the appropriate array based on the
+/// provided data type.
+/// @param array_iterator An iterator to consume column arrays.
+/// @param arrow_type the data type we're building the array for.
+std::shared_ptr<arrow::Array>
+make_arrow_array(arrow::ArrayVector::const_iterator& array_iterator,
+                 std::shared_ptr<arrow::DataType> arrow_type) {
+  auto f = detail::overload{
+    [&](const std::shared_ptr<arrow::Array>& a,
+        const std::shared_ptr<arrow::DataType>& at)
+      -> std::shared_ptr<arrow::Array> {
+      VAST_ASSERT(a->type()->Equals(at));
+      return a;
+    },
+    [&](const std::shared_ptr<arrow::StringArray>& pa,
+        const std::shared_ptr<pattern_extension_type>& pt)
+      -> std::shared_ptr<arrow::Array> {
+      return std::make_shared<pattern_array>(pt, pa);
+    },
+    [&](const std::shared_ptr<arrow::FixedSizeBinaryArray>& ba,
+        const std::shared_ptr<address_extension_type>& at)
+      -> std::shared_ptr<arrow::Array> {
+      VAST_ASSERT(ba->byte_width() == 16);
+      return std::make_shared<address_array>(at, ba);
+    },
+    [&](const std::shared_ptr<arrow::FixedSizeBinaryArray>& ba,
+        const std::shared_ptr<subnet_extension_type>&)
+      -> std::shared_ptr<arrow::Array> {
+      VAST_ASSERT(ba->byte_width() == 17);
+      auto subnet_array_data = convert_subnet_array(ba);
+      // TODO: the type is not actually correct, we have
+      // left:  struct<address: fixed_size_binary[16], length: uint8>
+      // right: struct<address: extension<vast.address>, length: uint8>
+      // so we can't create the subnet_array either (arrow asserts storage
+      // type) this gets "fixed" by arrow when re-reading the record batch
+      // in, but the record batch under construction exposes the wrong type
+      // information. to fix this, we require proper AddressBuilder /
+      // SubnetBuilder IIUC
```

Comment:
A better way to write this would be to have this overload first split the 17-byte-width Array into a 16-byte-width Array and the UInt8Array, and then to recursively call this function again with he 16-byte-width Array to convert it to an address type. This would need to be done regardless of whether or not we have a custom builder for addresses and subnets.

Until we have said builder, I think you must also call `{subnet, address}_extension_type::MakeArray` on the underlying `arrow::ArrayData` of the extension arrays.

Reply:
I would propose to wait for your changes and rebase / fix in this branch when the builders are available.